core/invoker/src/main/scala/whisk/core/containerpool/docker/ProcessRunner.scala

```diff
@@ -35,19 +37,30 @@ trait ProcessRunner {
    * @param args command to be run including arguments
    * @return a future completing according to the command's exit code
    */
-  protected def executeProcess(args: String*)(implicit ec: ExecutionContext) =
-    Future(blocking {
-      val out = new mutable.ListBuffer[String]
-      val err = new mutable.ListBuffer[String]
-      val exitCode = args ! ProcessLogger(o => out += o, e => err += e)
-
-      (exitCode, out.mkString("\n"), err.mkString("\n"))
-    }).flatMap {
-      case (0, stdout, _) =>
-        Future.successful(stdout)
-      case (code, stdout, stderr) =>
-        Future.failed(ProcessRunningException(code, stdout, stderr))
-    }
+  protected def executeProcess(args: String*)(implicit ec: ExecutionContext): Future[String] = {
+    val promise = Promise[String]()
+    val pb: NuProcessBuilder = new NuProcessBuilder(args.asJava)
```

Comment:
what's the advantage of this vs using a custom runner - like in TestUtils.runCmd?

Reply:
As said in the commit message, in a nutshell:

* Use epoll to detect finished processes --> thread consumption reduced
* Use a centralized thread pool for stream consuming. The current (scala.sys.Process) runner starts one unnamed Thread **per command** to do this. These are barebone threads, not manageable by any pool whatsoever.

Point for NuProcess: Akka uses it in its akka-stream contrib module for async processes.